File: templates/basic/dep.py
#! /usr/bin/env python

import argparse
import os

import sys

import subprocess
from os.path import expanduser


class Options:
    def __init__(self,
                 meta_config,
                 meta_path,
                 dev,
                 luarocks_config,
                 luarocks_tree):
        self.meta_config = meta_config
        self.meta_path = meta_path
        self.dev = dev
        self.luarocks_config = luarocks_config
        self.luarocks_tree = luarocks_tree


def ensure_tarantool_rocks_repo(luarocks_config):
    luarocks_config_dir = os.path.dirname(luarocks_config)
    if not os.path.isdir(luarocks_config_dir):
        os.mkdir(luarocks_config_dir)

    if os.path.exists(luarocks_config):
        with open(luarocks_config, 'r') as f:
            f_contents = f.read()
            if 'rocks.tarantool.org' in f_contents:
                return

    with open(luarocks_config, 'a') as f:
        f.write('\nrocks_servers = {[[http://rocks.tarantool.org/]]}\n')


def luarocks_install(dep, local=True, tree=None):
    cmd = ['luarocks', 'install']
    if tree:
        cmd.append("--tree={0}".format(tree))
    elif local:
        cmd.append('--local')
    cmd.append(dep)

    process = subprocess.Popen(cmd)
    exit_code = process.wait()
    
    assert exit_code == 0, "{0} failed".format(cmd)

    print('{0} finished'.format(cmd, exit_code))


def run(opts):
    app_name = opts.meta_config.get('name')
    assert app_name, 'name must be defined'

    # noinspection PyRedeclaration
    def fprint(s):
        print('[{0}] {1}'.format(app_name, s))

    fprint('Installing dependencies...')

    general_deps = opts.meta_config.get('deps', [])
    tnt_deps = opts.meta_config.get('tntdeps', [])

    if not general_deps and not tnt_deps:
        fprint('Nothing to install')

    if tnt_deps:
        ensure_tarantool_rocks_repo(opts.luarocks_config)
        for dep in tnt_deps:
            luarocks_install(dep, local=opts.dev, tree=opts.luarocks_tree)
            fprint('Installed tntdep: {0}'.format(dep))

    if general_deps:
        for dep in general_deps:
            luarocks_install(dep, local=opts.dev, tree=opts.luarocks_tree)
            fprint('Installed dep: {0}'.format(dep))

    return 0


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--meta-file', help='meta.json file with tarantool deps', required=True)
    parser.add_argument('--dev', help='pass True if this is dev', type=bool, default=False)
    parser.add_argument('--json', help='use json meta file instead of yaml', type=bool, default=False)
    parser.add_argument('--luarocks-config',
                        help='path to luarocks config file',
                        type=str,
                        default=expanduser('~/.luarocks/config.lua'))
    parser.add_argument('--luarocks-tree',
                        help='path to luarocks installation tree',
                        type=str,
                        default=None)
    args = parser.parse_args()

    meta_file = os.path.abspath(args.meta_file)
    with open(meta_file, 'r') as f:
        if args.json:
            import json
            meta_config = json.load(f, encoding='utf-8')
        else:
            import yaml
            meta_config = yaml.load(f)

    opts = Options(meta_config=meta_config,
                   meta_path=os.path.dirname(meta_file),
                   dev=args.dev,
                   luarocks_config=args.luarocks_config,
                   luarocks_tree=args.luarocks_tree)
    return run(opts)

if __name__ == "__main__":
    sys.exit(main())

File: templates/basic/template/dep.lua
#! /usr/bin/env tarantool

assert(_TARANTOOL ~= nil, 'dep.lua must be run by tarantool')

local fio = require 'fio'
local yaml = require 'yaml'

local cfg


local function fprint(f, ...)
    print(string.format('[%s] ' .. f, cfg.name, ...))
end


local function _is_dir(path)
    if _TARANTOOL >= "1.9" then
        return fio.path.is_dir(path)
    end

    local fs = fio.stat(path)
    return fs ~= nil and fs:is_dir() or false
end

local function _exists(path)
    if _TARANTOOL >= "1.9" then
        return fio.path.exists(path)
    end
    return fio.stat(path) ~= nil
end

local function _string_split(s, sep)
    sep = sep or ','
    local parts = {}
    for word in string.gmatch(s, string.format('([^%s]+)', sep)) do
        table.insert(parts, word)
    end
    return parts
end

local function ensure_rocksservers(path)
    local dir = fio.dirname(path)
    if not _is_dir(dir) then
        fio.mkdir(dir)
    end

    if _exists(path) then
        local f = fio.open(path)
        local data = f:read(f:stat().size)
        f:close()
        if data:match('rocks%.tarantool%.org') and data:match('rocks%.moonscript%.org') then
            fprint('Already have proper rocks servers')
            return
        end
    else
        local directory = fio.dirname(path)
        if not fio.mktree(directory) then
            fprint('Error while creating %s: %s', directory, errno.strerror())
            os.exit(1)
        end
    end
    fprint("Patch %s with proper rocks servers", path)
    local fh = fio.open(path, {'O_CREAT', 'O_APPEND', 'O_RDWR'}, 0664)
    fh:write('\nrocks_servers = {[[http://rocks.tarantool.org/]], [[https://rocks.moonscript.org]]}\n')
    fh:close()
end


local function execute(cmd)
    local raw_cmd = table.concat(cmd, ' ')
    fprint("%s...", raw_cmd)
    local res = os.execute(raw_cmd)
    if res ~= 0 then
        error(string.format('[%s] %s failed', cfg.name, raw_cmd))
    end
end


local function cmd_luarocks(subcommand, dep, tree)
    assert(subcommand ~= nil, 'subcommand is required')
    assert(dep ~= nil, 'dep is required')

    local cmd = {'luarocks', subcommand, dep}
    if tree then
        table.insert(cmd, '--tree='..tree)
    end
    return execute(cmd)
end


local function cmd_tarantoolctl(subcommand, dep, tree)
    assert(subcommand ~= nil, 'subcommand is required')
    assert(dep ~= nil, 'dep is required')

    local cmd = {'tarantoolctl', 'rocks', subcommand, dep}
    return execute(cmd)
end


local function _gencmd(command, subcommand)
    return function(dep, tree)
        return command(subcommand, dep, tree)
    end
end

local luarocks_install = _gencmd(cmd_luarocks, 'install')
local luarocks_remove = _gencmd(cmd_luarocks, 'remove')
local luarocks_make = _gencmd(cmd_luarocks, 'make')
local tarantoolctl_install = _gencmd(cmd_tarantoolctl, 'install')


local function main()
    print('Tarantool version: ' .. _TARANTOOL)

    local luaroot = debug.getinfo(1, 'S')
    local source = fio.abspath(luaroot.source:match('^@(.+)'))
    local appname = fio.basename(fio.dirname(source))

    local args = {
        ['--luarocks-config'] = fio.pathjoin(os.getenv('HOME'), '.luarocks', 'config.lua'),
        ['--meta-file']       = './meta.yaml',
        ['--tree']            = '.rocks',
        ['--only']            = '',
    }

    for i = 1,#arg/2 do args[ arg[i*2-1] ] = arg[i*2] end

    local meta_path = args['--meta-file']
    assert(meta_path ~= '', 'meta file is required')

    print('Using the following options:\n' .. yaml.encode(args))

    local meta_file = fio.open(fio.abspath(meta_path))
    local metatext = meta_file:read(meta_file:stat().size)
    local tree = fio.abspath(args['--tree'])
    local only_sections
    if args['--only'] ~= '' then
        only_sections = {}
        for _, s in ipairs(_string_split(args['--only'])) do
            only_sections[s] = true
        end
    end

    cfg = metatext:match('^%s*%{') and require 'json'.decode(metatext) or yaml.decode(metatext)

    cfg.name = cfg.name or appname
    assert(cfg.name, 'Name must be defined')

    ensure_rocksservers(args['--luarocks-config'])

    fprint('Installing dependencies...')
    local deps = cfg.deps or {}
    local tnt_deps = cfg.tnt_deps or cfg.tntdeps or {}
    local local_deps = cfg.local_deps or cfg.localdeps or {}

    if only_sections == nil or only_sections.deps then
        for _, dep in ipairs(deps) do
            fprint("Installing dep '%s'", dep)
            luarocks_install(dep, tree)
            fprint("Installed dep '%s'\n\n", dep)
        end
    end

    if only_sections == nil or only_sections.tntdeps or only_sections.tnt_deps then
        for _, dep in ipairs(tnt_deps) do
            fprint("Installing tarantool dep '%s'", dep)
            tarantoolctl_install(dep, tree)
            fprint("Installed tarantool dep '%s'\n\n", dep)
        end
    end

    if only_sections == nil or only_sections.localdeps or only_sections.local_deps then
        local cwd = fio.cwd()
        for _, dep in ipairs(local_deps) do
            local dep_root
            local dep_info = _string_split(dep, ':')
            if #dep_info > 1 then
                dep = dep_info[1]
                dep_root = dep_info[2]
            else
                dep_root = fio.dirname(dep)
            end
            fprint("Installing local dep '%s' with root at '%s'", dep, dep_root)
            dep = fio.abspath(dep)
            dep_root = fio.abspath(dep_root)

            fio.chdir(dep_root) -- local rocks must be installed from within the project root
            local ok, res = pcall(luarocks_remove, dep, tree)
            if not ok then
                fprint(res)
            end

            luarocks_make(dep, tree)
            fio.chdir(cwd)
            fprint("Installed local dep '%s'\n\n", dep)
        end
    end
    fprint('Done.')
end

xpcall(main, function(err)
    print(err .. '\n' .. debug.traceback())
    os.exit(1)
end)
